colossalai/utils/model/lazy_init_context.py

```diff
@@ -187,71 +189,61 @@ def lazy_init_parameters(self, model: torch.nn.Module, device='cpu', call_back:
         for name, buffer in model.named_buffers():
             param_id_to_name[id(buffer)] = name
 
-        assert model in self.module_rebuild_dict, 'We only support rebuild modules which intercepted during initializing by us.'
-
-        def _process_arg(arg):
-            """
-            Process args recursively. If arg is a torch.nn.Module instance in module_rebuild_dict, 
-            we need to rebuild it with real parameters. If arg is a tuple or list, we will process
-            the element of arg with this function again.
-            """
-            if torch.is_tensor(arg):
-                tensor_id = id(arg)
-                if tensor_id in param_id_to_name:
-                    arg = _replace_meta_param_with_real_param(arg)
-
-            elif isinstance(arg, torch.nn.Module):
-                if arg in self.module_rebuild_dict:
-                    arg = self.lazy_init_parameters(model=arg, device=device, call_back=call_back)
-
-            elif isinstance(arg, (tuple, list)):
-                rst_list = []
-                for element in arg:
-                    processed_element = _process_arg(element)
-                    rst_list.append(processed_element)
-                arg = rst_list
-            return arg
-
-        def _replace_meta_param_with_real_param(meta_param):
-            if meta_param.device != 'meta':
-                return meta_param
-            tensor_id = id(meta_param)
+        def _init_recurively(module):
+            for mod in module.children():
+                _init_recurively(mod)
+
+            for param in module.parameters(recurse=False):
+                _convert_to_real_tensor(param, module, is_buffer=False)
+
+            for buf in module.buffers(recurse=False):
+                _convert_to_real_tensor(buf, module, is_buffer=True)
+
+        def _convert_to_real_tensor(tensor, module, is_buffer):
+            if not tensor.is_meta:
+                return tensor
+            tensor_id = id(tensor)
             param_full_name = param_id_to_name[tensor_id]
-            real_param = torch.empty_like(meta_param, dtype=meta_param.dtype, device=device)
-            real_param = ColoParameter(real_param, requires_grad=meta_param.requires_grad)
+            real_tensor = torch.empty_like(tensor, dtype=tensor.dtype, device=device)
 
-            if '.' in param_full_name:
-                submodule_name, param_name = param_full_name.rsplit('.', 1)
-                submodule = model.get_submodule(submodule_name)
+            # look for initialization function
+            if tensor in self._intercepted_nn_init_func_cache:
+                init_func, args, kwargs = self._intercepted_nn_init_func_cache[tensor][-1]
             else:
-                submodule = model
-                param_name = param_full_name
-            setattr(submodule, param_name, real_param)
+                raise RuntimeError(f"{param_full_name} is not associated with a initialization function")
```

Comment:
Some parameters may be initialized in the way like 'self.bias = nn.Parameter(torch.zeros(3,3))'

Reply:
This is ok. All `torch.nn` modules call `reset_parameters` in `__init__`.

Reply:
https://github.com/huggingface/transformers/blob/d4ebd4e112034b4a429ab7f813d7e168e7bb63c3/src/transformers/pytorch_utils.py#L88

Reply:
Ok I see.